### PR TITLE
digital: add correlate access code tagging block with soft inputs

### PR DESCRIPTION
**Pull request by nowls**

### CODE REVIEW

**Comment by nowls** [comment]
The only issue I foresee is with GRC integration. I removed the XML file in favor of one that can switch IO types between byte and float. Please let me know if you think that would break anything. I have a stand-alone float version I could submit.

File: gr-digital/lib/correlate_access_code_tag_ff_impl.cc [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+#include <gnuradio/math.h>
+#include <stdexcept>
+#include <volk/volk.h>
+#include <cstdio>
+#include <iostream>
+
+namespace gr {
+  namespace digital {
+
+#define VERBOSE 0
```
Comment (jmcorgan) [inline gr-digital/lib/correlate_access_code_tag_ff_impl.cc]: We have the logging system for this now.  You can replace the debug output to stdout/stederr with GR_LOG_DEBUG calls, and the logging system will direct those messages as configured by the user.

File: gr-digital/lib/correlate_access_code_tag_ff_impl.cc [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    correlate_access_code_tag_ff_impl::correlate_access_code_tag_ff_impl(
+        const std::string &access_code, int threshold, const std::string &tag_name)
+      : sync_block("correlate_access_code_tag_ff",
+		      io_signature::make(1, 1, sizeof(float)),
+		      io_signature::make(1, 1, sizeof(float))),
+	d_data_reg(0), d_mask(0),
+	d_threshold(threshold), d_len(0)
+    {
+      if(!set_access_code(access_code)) {
+	throw std::out_of_range ("access_code is > 64 bits");
```
Comment (jmcorgan) [inline gr-digital/lib/correlate_access_code_tag_ff_impl.cc]: This is a chance to do a GR_LOG_ERROR call before throwing.

File: gr-digital/lib/correlate_access_code_tag_ff_impl.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+      // set len bottom bits to 1.
+      d_mask = ((~0ULL) >> (64 - d_len));
+
+      d_access_code = 0;
+      for(unsigned i=0; i < d_len; i++){
+        d_access_code = (d_access_code << 1) | (access_code[i] & 1);
+      }
+
+      if(VERBOSE) {
```
Comment (jmcorgan) [inline gr-digital/lib/correlate_access_code_tag_ff_impl.cc]: As noted above.

File: gr-digital/lib/correlate_access_code_tag_ff_impl.cc [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	uint64_t wrong_bits = 0;
+	uint64_t nwrong = d_threshold+1;
+
+	wrong_bits  = (d_data_reg ^ d_access_code) & d_mask;
+	volk_64u_popcnt(&nwrong, wrong_bits);
+
+	// shift in new data
+	d_data_reg = (d_data_reg << 1) | (gr::branchless_binary_slicer(in[i]) & 0x1);
+	if(nwrong <= d_threshold) {
+	  if(VERBOSE)
```
Comment (jmcorgan) [inline gr-digital/lib/correlate_access_code_tag_ff_impl.cc]: As noted above.

**Comment by jmcorgan** [comment]
Hey @nowls, let me know when you think you'll have time to resume work on this, just so I can plan. As a new block it's not going to go stale very quickly but I like to keep these to a minimum.  Thanks.

**Comment by nowls** [comment]
I followed the lead of the other blocks when writing the new one. It looks like similar changes can be applied to the other correlate_access_code_* blocks. I can take care of these this week.

**Comment by jmcorgan** [comment]
Yeah, while we have bad habits strewn all over the existing code base, I try to keep them from proliferating into new code.  

It reinforces the notion that existing code is documentation for new code.

As an aside for anyone reading this, here's a perfect way to make small, incremental changes to GNU Radio--fix cases where we are outputting info to stdout/stderr and use the logging system instead.

**Comment by nowls** [comment]
Would you be comfortable accepting this PR as-is, and I can make a separate PR with updates to the logging for all correlate_access_code_* blocks to use GR_LOGGER? I think that makes more sense logically but it's up to you.

**Comment by jmcorgan** [comment]
Sure, we can do it that way.

**Comment by nowls** [comment]
Ok, great. I'll make the other changes this week. Thanks!

**Pull request merged**